controllers/C_Document.class.php

```diff
@@ -795,6 +801,15 @@ public function retrieve_action(string $patient_id = null, $document_id, $as_fil
         if (file_exists($temp_url)) {
             $url = $temp_url;
         }
+        //fire a remote call to see if the file is stored somewhere else
+        $s3Key = explode("//", $temp_url); //split the url to get the s3 key
+        $retrieveOffsiteDocument = new PatientRetrieveOffsiteDocument("/" . $s3Key[1]);
+        $this->eventDispatcher->dispatch($retrieveOffsiteDocument, PatientRetrieveOffsiteDocument::REMOTE_DOCUMENT_LOCATION);
+        //this is for the s3 bucket module. If the file is not found locally, it will be found remotely
+        if ($GLOBALS['presignedUrl']) {
```

Comment:
Can't you get back the dispatched event and use that instead of this global presignedUrl?  This feels backwards or I'm misunderstanding something as you should be able to get flow back from your event and use that in the if statement instead of relying on some global setting value.

Reply:
Looking at your event, why don't you pass in your $temp_url into your event, do any s3 specific stuff in your module, and have a boolean flag on the event for hasExternalUrlHandler() which you can then use on line 809.  Your event could then have a getExternalUrl that you populate in your event and then use it on line 810. That makes the code generic instead of relying on this global presignedUrl.   People can then use the same stuff to add an azure backing, or whatever else they need. At that point this would look fine to me and I'd be comfortable bringing it in.

Reply:
I'm hoping this is a pretty minor change if you are understanding what I'm saying.  I realize you've been waiting on this for a while, but I will make sure to review it and merge it in this week @juggernautsei as I know you've been waiting for a while.

Reply:
Not simple. The pre-signed URL that is generated by AWS SDK looks like this.

      (https://openemrtestbucket.s3.amazonaws.com//oe-66207cad3b04d.pdf?X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Security-Token=IQoJb3JpZ2luX2VjEKr%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJHMEUCIQDvGQQI0bqtkPtoRD9H%2FO4AldN6u8%2B4gCB0AajswSLbywIgMUx6sgjCvqSm2FA6hbF72euF7%2FF3NikDsgkAlMwpWYEqvAUI4%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARADGgw2NDU2NDg5NDM1NjMiDI0w3VKAn9NBwlvdaSqQBfDjo9G%2FlSGWJWYssS8uxQ2gmL9eT9%2F3P65e19TegKwUATNNUi9wRw6DQOl%2Fwo97HZeJvhrF0Ps%2Fp6Bdz2DVkIiviYOaxACqAiqtCbpktvtekefAkiJcdPio8%2FfIV%2Fo0RD%2BSYTJK0GubhnQrqSwvX%2FMznWLX5U%2F7IEtuJW0Mgtkx5uYLbU1IxHFI%2FPMIk6WZLt1GRBG5bSy5KPFpbxHN%2FlJezrUcsAnxU%2BBBZFQhtVGDYXNTNaljG9XEnweQPADgrIuCZM9tIqIL9kMMA9k6krvUAPBfMQ6rXrTQDrmfxBU0nsoWGeZaWe4zTNk596m9K4ex1xFKRZdeX8MFe2ASu1Zinn4Cb7P42y%2BKQCFEhu5t%2FAlB3eqwLsBn%2FBrkD8VXG9zXel9L5Jczx3Ev2O%2BfI2IHLOTlV3HWdHTr3T7EbacdGXqu5FmDcrlelJT7U0R226OdVqV9eDT7FBuvKYCGCiAUYO%2Bom7RYjS2sy62Rcll64cPDr6rrIpMzdU5R2nYh5Nl6noepyCZZaVC3%2BWz1j1eDuRWM%2FUEyy1J9WKvTfmkzVOXNC81NQNGYXSCrQqVDGCjlBdyPP%2B0GGi6wmLp0kO%2FXgUHG7CTNnm5alz4%2B6wY%2FJZhBSX%2BvhnuqywNhNAYG%2BDQ9ijVFBu75OmGgyY3%2BVNHcYBHnGf2sV3ekbgyq2AMfa4CTfGD0QnjmOZeyhjFF2i49BJkusKJiSPD7vnrx5JZWDBkPqyGJ9HEgnJVUfFeaoNkLKeRNxDa7cCttgHJ6tIlWGk3QwS8XaqY7iTip24bdfLqiUGt40iDHPZkRix%2F%2FNB8AGdevMMg2Xc2GTp20pt86jTvUp%2B%2B76emYd6L8xE34yy4QmgkjEq6JzHYgvHIcMKz9gbEGOrEBVqKMVK8ijnzGEmqtjlQeeGDG2WAw2N4bOGaKGw%2BlEReU1b3jqOjkZqMFFBp%2BDHxKsyz3N54tSds%2BQf9KI4tr%2F2d%2BGUZDvpAEZUDQ6IJRYMFuyQmOoAA935yBAibmyb0j%2FO4DEDPR4B3ni%2FAfh4eZa5er8haz6Xw0XoUNjD6YyNheWtmCNp4MVdah43O8NkAUgAb48rCi0Znk1fk4TEs%2F8TJA1vxO3x679nESsbp123zb&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIAZMU5NBHFSH2LLZWP%2F20240418%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240418T022042Z&X-Amz-SignedHeaders=host&X-Amz-Expires=1200&X-Amz-Signature=a8f8e9782d32585e2e55284c0cb83e6ac0082ac49a20828775b1ad50c719c8e7)

When I return it. All I get is the file name in the S3 bucket.

    public function retrieveFromS3(PatientRetrieveOffsiteDocument $event)
    {
        $url = $event->url;
        $s3model = new S3Model();
        $credentials = new Credentials($s3model);
        $s3Client = $credentials->authenticateCredentials();
        $getBucket = $s3model->getAllCredentials();
        $cmd = $s3Client->getCommand('GetObject', [
            'Bucket' => $getBucket['s3_bucket'],
            'Key' => $url
        ]);
        $request = $s3Client->createPresignedRequest($cmd, '+20 minutes');
        $presignedUrl = (string) $request->getUri();
        file_put_contents('/var/www/html/quest/obj_presigned.log', $presignedUrl . PHP_EOL, FILE_APPEND);
        return $presignedUrl;
    }

This is the returned object.

    OpenEMR\Events\PatientDocuments\PatientRetrieveOffsiteDocument Object
    (
        [propagationStopped:Symfony\Contracts\EventDispatcher\Event:private] => 
        [url] => /oe-66207cad3b04d.pdf
    )

When I put it in a Global. I can use it but simply returning the URL does not work.

Reply:
https://dnl7t01l0fo05.cloudfront.net/wp-content/uploads/2015/12/jack-russell-gif-1.gif